src/linear_algebra/basic.lean

```diff
@@ -127,6 +127,19 @@ ext $ assume b, rfl
   p.subtype.comp (cod_restrict p f h) = f :=
 ext $ assume b, rfl
 
+/-- Restrict domain and codomain of an endomorphism. -/
+def restrict (f : M →ₗ[R] M) (p : submodule R M) (hf : ∀ x ∈ p, f x ∈ p) : p →ₗ[R] p :=
+{ to_fun := λ x, ⟨f x, hf x.1 x.2⟩,
+  map_add' := begin intros, apply set_coe.ext, simp end,
+  map_smul' := begin intros, apply set_coe.ext, simp end }
+
+lemma restrict_apply
+  {f : M →ₗ[R] M} {p : submodule R M} (hf : ∀ x ∈ p, f x ∈ p) (x : p) :
+  f.restrict p hf x = ⟨f x, hf x.1 x.2⟩ := rfl
+
+lemma subtype_comp_restrict {f : M →ₗ[R] M} {p : submodule R M} (hf : ∀ x ∈ p, f x ∈ p) :
+  p.subtype.comp (f.restrict p hf) = f.comp p.subtype := rfl
```

Comment:
`f.comp p.subtype` is also known as `f.dom_restrict p`. On the other hand, it looks like `dom_restrict` is only used once after its definition, so your version might have better library support. (In that case, the ideal solution is to improve library support for `dom_restrict`, but no need for that in this PR.)
```suggestion
  p.subtype.comp (f.restrict p hf) = f.dom_restrict p := rfl
```